controllers/space/space_controller.go

```diff
@@ -0,0 +1,343 @@
+	// so the Space can be in `ready` status as well
+	requeue, err := r.ensureNSTemplateSet(logger, space)
+	return ctrl.Result{Requeue: requeue}, err
+}
+
+// setFinalizers sets the finalizers for Space
+func (r *Reconciler) addFinalizer(logger logr.Logger, space *toolchainv1alpha1.Space) error {
+	// Add the finalizer if it is not present
+	if !util.HasFinalizer(space, toolchainv1alpha1.FinalizerName) {
+		logger.Info("adding finalizer on Space")
+		util.AddFinalizer(space, toolchainv1alpha1.FinalizerName)
+		if err := r.Client.Update(context.TODO(), space); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ensureNSTemplateSet creates the NSTemplateSet on the target member cluster if it does not exist,
+// and updates the space's status accordingly.
+// returns `true` after creating the NSTemplateSet and until it is `ready`
+func (r *Reconciler) ensureNSTemplateSet(logger logr.Logger, space *toolchainv1alpha1.Space) (bool, error) {
+	if space.Spec.TargetCluster == "" {
+		return false, r.setStatusProvisioningFailed(logger, space, fmt.Errorf("unspecified target member cluster"))
+	}
+	memberCluster, found := r.MemberClusters[space.Spec.TargetCluster]
+	if !found {
+		return false, r.setStatusProvisioningFailed(logger, space, fmt.Errorf("unknown target member cluster '%s'", space.Spec.TargetCluster))
+	}
+	name := space.Name // NSTemplateSet will have the same name as the Space
+	logger = logger.WithValues("target_member_cluster", space.Spec.TargetCluster, "name", name)
+	// create if not found
+	nsTmplSet := &toolchainv1alpha1.NSTemplateSet{}
+	if err := memberCluster.Client.Get(context.TODO(), types.NamespacedName{
+		Namespace: memberCluster.OperatorNamespace,
+		Name:      space.Name,
+	}, nsTmplSet); err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("creating NSTemplateSet on target member cluster")
+			if err := r.setStatusProvisioning(space); err != nil {
+				return false, err
+			}
+			if nsTmplSet, err = r.newNSTemplateSet(memberCluster.OperatorNamespace, space); err != nil {
+				return false, r.setStatusProvisioningFailed(logger, space, err)
+			}
+
+			if err := memberCluster.Client.Create(context.TODO(), nsTmplSet); err != nil {
+				logger.Error(err, "failed to create NSTemplateSet on target member cluster")
+				return false, r.setStatusNSTemplateSetCreationFailed(logger, space, err)
+			}
+			logger.Info("NSTemplateSet created on target member cluster")
+			return false, nil
+		}
+		return false, r.setStatusNSTemplateSetCreationFailed(logger, space, err)
+	}
+	logger.Info("NSTemplateSet already exists")
+
+	readyCond, found := condition.FindConditionByType(nsTmplSet.Status.Conditions, toolchainv1alpha1.ConditionReady)
+	if !found || readyCond.Status != corev1.ConditionTrue {
+		logger.Info("NSTemplateSet is not ready", "ready-condition", readyCond)
+		return true, nil // here we need to explicitly requeue since the controller doesn't watch the NSTemplateSetStatus
+	}
```

Comment:
We need to replicate the unready conditions. I mentioned it in the last review https://github.com/codeready-toolchain/host-operator/pull/548#discussion_r759338529
 
in addition - you wrote 

> // here we need to explicitly requeue since the controller doesn't watch the NSTemplateSetStatus

but the controller watches the changes in NSTemplatSet.status - we don't use any predicate for it. Or am I missing something?

Reply:
I'm confused because your comment above ended with 

> Anyway, if the condition is either not found or unknown, then we should not replicate it to Space CR.
In UserAccount controller, we don't replicate the status if the UserAccount ready condition is already set to false to just not overload the synchronization mechanism. **We could technically do the same also here, but it's not super needed**

so, am I supposed to do it or not? Besides, we don't have the same kind of sync mechanism as in the UserAccount, so what's the point here?

Reply:
> but the controller watches the changes in NSTemplatSet.status - we don't use any predicate for it. Or am I missing something?

this was addressed in https://github.com/codeready-toolchain/host-operator/pull/548/commits/47eab1f38e323375dd1afa43d1b39fa6de741d3e

Reply:
@xcoulon We need to replicate the status from NSTempalteSet CR to Space CR so we know what's going on with the NSTemplateSet. 
By saying 

> Anyway, if the condition is either not found or unknown, then we should not replicate it to Space CR.
In UserAccount controller, we don't replicate the status if the UserAccount ready condition is already set to false to just not overload the synchronization mechanism. **We could technically do the same also here, but it's not super needed**

I was referring to this particular case https://github.com/codeready-toolchain/member-operator/blob/master/controllers/useraccount/useraccount_controller.go#L404-L408 - in UserAccount we don't replicate the status for the case when the ready condition is already set to false and NSTemplateSet doesn't provide any message. 
Since we don't have to deal with the synchronization mechanism here, we can skip that part and replicate the status for all cases when the condition is not ready https://github.com/codeready-toolchain/member-operator/blob/master/controllers/useraccount/useraccount_controller.go#L410-L416

Reply:
ok, thanks for the clarification. I will fix that

Reply:
done in ac77948, hopefully!